### PR TITLE
Add support for dlang's std.regex

### PR DESCRIPTION
**Pull request by Mai-Lapyst**
This is a experimental draft PR where support for dlang's `std.regex` is added.

### CODE REVIEW

**Comment by BurntSushi** [comment]
Woohoo! I haven't given this a deep dive review yet, but the formatting of the D source file is kind of nutty:

formatting

Can you remove tabs (use 4-space indents) and wrap lines to 79 columns (inclusive)?

Also it looks like the D engine is missing from some curated benchmarks. For example, the `04-ruff` benchmarks.

**Comment by BurntSushi** [comment]
These are the results I get relative to Rust's regex engine:

```
$ rebar cmp tmp/results.csv --intersection -M compile
[2023-10-07T21:53:27Z WARN  rebar::format::measurement] curated/13-noseyparker/single:dlang/std: skipping because of error: timeout: exceeded 10s
benchmark                                       dlang/std             rust/regex
---------                                       ---------             ----------
curated/01-literal/sherlock-en                  6.3 GB/s (5.15x)      32.3 GB/s (1.00x)
curated/01-literal/sherlock-casei-en            1745.9 MB/s (5.68x)   9.7 GB/s (1.00x)
curated/01-literal/sherlock-ru                  814.0 MB/s (38.92x)   30.9 GB/s (1.00x)
curated/01-literal/sherlock-casei-ru            1933.4 MB/s (4.49x)   8.5 GB/s (1.00x)
curated/01-literal/sherlock-zh                  2.5 GB/s (15.80x)     40.0 GB/s (1.00x)
curated/02-literal-alternate/sherlock-en        686.1 MB/s (17.02x)   11.4 GB/s (1.00x)
curated/02-literal-alternate/sherlock-casei-en  680.6 MB/s (4.47x)    3.0 GB/s (1.00x)
curated/02-literal-alternate/sherlock-ru        979.0 MB/s (6.64x)    6.3 GB/s (1.00x)
curated/02-literal-alternate/sherlock-casei-ru  851.0 MB/s (1.78x)    1512.9 MB/s (1.00x)
curated/02-literal-alternate/sherlock-zh        1860.0 MB/s (7.26x)   13.2 GB/s (1.00x)
curated/03-date/ascii                           404.7 KB/s (418.51x)  165.4 MB/s (1.00x)
curated/03-date/unicode                         401.7 KB/s (418.90x)  164.3 MB/s (1.00x)
curated/06-cloud-flare-redos/original           12.6 MB/s (45.51x)    573.3 MB/s (1.00x)
curated/06-cloud-flare-redos/simplified-short   17.4 MB/s (119.15x)   2.0 GB/s (1.00x)
curated/06-cloud-flare-redos/simplified-long    20.7 MB/s (4273.15x)  86.2 GB/s (1.00x)
curated/08-words/all-english                    17.5 MB/s (6.88x)     120.2 MB/s (1.00x)
curated/08-words/all-russian                    28.3 MB/s (1.00x)     18.8 MB/s (1.50x)
curated/08-words/long-english                   1212.3 MB/s (1.00x)   801.5 MB/s (1.51x)
curated/08-words/long-russian                   57.7 MB/s (1.00x)     32.1 MB/s (1.80x)
curated/09-aws-keys/full                        13.6 MB/s (130.81x)   1779.0 MB/s (1.00x)
curated/09-aws-keys/quick                       262.0 MB/s (7.21x)    1888.5 MB/s (1.00x)
curated/10-bounded-repeat/letters-en            179.3 MB/s (4.21x)    754.8 MB/s (1.00x)
curated/10-bounded-repeat/letters-ru            64.5 MB/s (10.20x)    658.3 MB/s (1.00x)
curated/10-bounded-repeat/context               125.6 MB/s (5.61x)    704.9 MB/s (1.00x)
curated/10-bounded-repeat/capitals              193.1 MB/s (4.28x)    825.6 MB/s (1.00x)
curated/12-dictionary/single                    30.7 MB/s (23.29x)    714.5 MB/s (1.00x)
curated/14-quadratic/1x                         673.0 KB/s (27.64x)   18.2 MB/s (1.00x)
curated/14-quadratic/2x                         358.6 KB/s (24.76x)   8.7 MB/s (1.00x)
curated/14-quadratic/10x                        76.1 KB/s (22.47x)    1710.1 KB/s (1.00x)
```

Do those look about right to you?

**Comment by BurntSushi** [comment]
The `compile` benchmark results look suspicious:

```
$ rebar cmp tmp/results.csv --intersection -m compile
[2023-10-07T22:03:02Z WARN  rebar::format::measurement] curated/13-noseyparker/single:dlang/std: skipping because of error: timeout: exceeded 10s
benchmark                                   dlang/std         rust/regex
---------                                   ---------         ----------
curated/03-date/compile-ascii               3.40us (1.00x)    1.48ms (435.29x)
curated/03-date/compile-unicode             3.40us (1.00x)    5.14ms (1511.76x)
curated/05-lexer-veryl/compile-single       800.00ns (1.00x)  271.42us (339.27x)
curated/07-unicode-character-data/compile   100.00ns (1.00x)  27.79us (277.90x)
curated/09-aws-keys/compile-full            200.00ns (1.00x)  86.22us (431.10x)
curated/09-aws-keys/compile-quick           100.00ns (1.00x)  14.43us (144.30x)
curated/10-bounded-repeat/compile-capitals  100.00ns (1.00x)  57.20us (572.00x)
curated/11-unstructured-to-json/compile     100.00ns (1.00x)  19.38us (193.80x)
curated/12-dictionary/compile-single        27.70us (1.00x)   7.54ms (272.20x)
curated/13-noseyparker/compile-single       5.00us (1.00x)    2.36ms (472.00x)
```

The fact that so many are clocking in at _exactly_ 100.00ns makes me think something is wrong. Does D's regex library cache compilation?

**Comment by Mai-Lapyst** [comment]
First of: thanks for the quick response!

> Can you remove tabs (use 4-space indents) and wrap lines to 79 columns (inclusive)?

Ofc, can do. Only opened the PR to show / gauge interest. Still needs some work.

> Also it looks like the D engine is missing from some curated benchmarks. For example, the 04-ruff benchmarks.

Yeah, some of them fail with a crash so I didnt added them there. Need to look further into them why tho.

> Do those look about right to you?

Yeah they look about right. Maybe using ldc/gdc instead will boost in production the numbers a notch, but dunno if thats a concern for the benchmarking tho. Espc bc it ment it needed to be run like 3 times, each for the different compiler implementation.

> The fact that so many are clocking in at exactly 100.00ns makes me think something is wrong. Does D's regex library cache compilation?

Yes it should cache some of them. That they are exactly 100ns is no surprise bc the default implementation of dlang's time measurement unfortunately runs on hnsecs. Also one aspects that I need to look more into.

**Comment by Mai-Lapyst** [comment]
After switching to an more precise clock for atleast the benchmarking times, it now looks a bit less odd:

```
$ cargo run cmp -m compile --intersection results.csv
    Finished dev [unoptimized + debuginfo] target(s) in 0.02s
     Running `target/debug/rebar cmp -m compile --intersection results.csv`
benchmark                                   dlang/std         rust/regex
---------                                   ---------         ----------
curated/03-date/compile-ascii               2.39us (1.00x)    2.40ms (1004.18x)
curated/03-date/compile-unicode             2.38us (1.00x)    9.10ms (3823.53x)
curated/05-lexer-veryl/compile-single       661.00ns (1.00x)  418.77us (633.54x)
curated/07-unicode-character-data/compile   181.00ns (1.00x)  43.62us (240.99x)
curated/09-aws-keys/compile-full            220.00ns (1.00x)  124.18us (564.45x)
curated/09-aws-keys/compile-quick           140.00ns (1.00x)  23.63us (168.79x)
curated/10-bounded-repeat/compile-capitals  140.00ns (1.00x)  104.90us (749.29x)
curated/11-unstructured-to-json/compile     160.00ns (1.00x)  29.57us (184.81x)
curated/12-dictionary/compile-single        19.25us (1.00x)   13.26ms (688.83x)
curated/13-noseyparker/compile-single       5.38us (1.00x)    3.77ms (700.74x)
```

**Comment by Mai-Lapyst** [comment]
Looked a bit into the missing benchmarks:

Missing benchmarks that cannot be supported with the current version of `std.regex`:
- no support for '(?i:)' syntax:
	- 04-ruff-noqa/real 
	- 04-ruff-noqa/tweaked
	- 04-ruff-noqa/compile-real
	- 10-bounded-repeat/compile-context
- no support for list of regex's (i.e. 'per-line = "pattern"'):
	- 05-lexer-veryl/multi
	- 12-dictionary/multi
	- 12-dictionary/compile-multi
	- 13-noseyparker/multi
	- 13-noseyparker/compile-multi

Missing benchmarks that need investigation:
- 05-lexer-veryl/single : mismatching count (5_491_200 / 124_800)
- 07-unicode-character-data/parse-line : mismatching count (523_860 / 558_784)
- 11-unstructured-to-json/extract : mismatching count (500 / 600)
- 13-noseyparker/single : timeout after 10s

**Comment by BurntSushi** [comment]
You've listed the license of this contribution as proprietary. I can't accept that. Do you consent to re-licensing it under the UNLICENSE?

**Comment by Mai-Lapyst** [comment]
> You've listed the license of this contribution as proprietary. I can't accept that. Do you consent to re-licensing it under the UNLICENSE?

Oh my bad. Ofcourse! Have also added a commit changing the relevant entry.

**Comment by BurntSushi** [comment]
> * no support for `(?i:)` syntax:
>
>   * 04-ruff-noqa/real
>   * 04-ruff-noqa/tweaked
>   * 04-ruff-noqa/compile-real
>   * 10-bounded-repeat/compile-context

I've modified these benchmarks to drop the use of inline flags. It's a pity
that ECMAScript has caught at least three of the engines in this benchmark
in its snare. Everything else supports inline flags.

> * no support for list of regex's (i.e. 'per-line = "pattern"'):
>
>   * 05-lexer-veryl/multi
>   * 12-dictionary/multi
>   * 12-dictionary/compile-multi
>   * 13-noseyparker/multi
>   * 13-noseyparker/compile-multi

Yeah that's expected. Only rust/regex and hyperscan support this.

> Other missing:
>
>     * 13-noseyparker/single : timeout after 10s; will remain exluded like many other engines

Yeah, this is a "tortue test."

> Missing benchmarks that need investigation:
>
> * 05-lexer-veryl/single : mismatching count (5_491_200 / 124_800)
> * 07-unicode-character-data/parse-line : mismatching count (523_860 / 558_784)
> * 11-unstructured-to-json/extract : mismatching count (500 / 600)

Yeah these match counts look very suspicious. Especially for these where
literally every other regex engine is in agreement. It's likely a bug in D's
regex engine or in your runner program. I've removed D from these benchmarks
(and the corresponding compilation benchmarks) and added a note saying why.

You can stop updating this PR now. I'll take it over from here. Specifically,
I've already done the following:

* Removed D from some benchmarks as mentioned above.
* Added D to the 04-ruff-noqa benchmarks by using `[Nn][Oo][Qq][Aa]` instead of
the `i` inline flag.
* Added D to the `10-bounded-repeat/compile-context` benchmark by replacing
`(?s:.)` with `[\S\s]`.
* Renamed the engine to `d/dmd/std-regex`. That way, `ldc` can be added later.
It is okay to do so. If `ldc` is always faster, then we can still keep `dmd`
around but drop it from the curated benchmark suite and just use `ldc`.
Basically, we should use whatever compiler people are using when they care
about perf.
* Cleaned up the engine definition in `engines.toml`.

Still left to do:

* Do a more thorough review of the runner program.
* Review D's regex engine source code for interesting things. For example,
if it is caching regexes and there is no way to turn it off, then I'll need
to remove the compile-time curated benchmarks for D. (As I've done for other
engines, such as `perl`.)

My PR supplanting this one is here (it builds off of yours):
https://github.com/BurntSushi/rebar/pull/11
